include/tracing.h

```diff
@@ -0,0 +1,129 @@
+/*
+ * Copyright (c) 2018 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#ifndef _KERNEL_TRACE_H
+#define _KERNEL_TRACE_H
+
+#include <kernel.h>
+
+#define SYS_TRACE_ID_OFFSET                  (32u)
+
+#define SYS_TRACE_ID_MUTEX_INIT              (1u + SYS_TRACE_ID_OFFSET)
+#define SYS_TRACE_ID_MUTEX_UNLOCK            (2u + SYS_TRACE_ID_OFFSET)
+#define SYS_TRACE_ID_MUTEX_LOCK              (3u + SYS_TRACE_ID_OFFSET)
+#define SYS_TRACE_ID_SEMA_INIT               (4u + SYS_TRACE_ID_OFFSET)
+#define SYS_TRACE_ID_SEMA_GIVE               (5u + SYS_TRACE_ID_OFFSET)
+#define SYS_TRACE_ID_SEMA_TAKE               (6u + SYS_TRACE_ID_OFFSET)
+
+#ifdef CONFIG_TRACING
+
+void sys_trace_thread_switched_out(void);
+
+void sys_trace_thread_switched_in(void);
+
+void sys_trace_thread_priority_get(struct k_thread *thread);
+
+void sys_trace_thread_priority_set(struct k_thread *thread);
+
+void sys_trace_thread_create(struct k_thread *thread);
+
+void sys_trace_thread_abort(struct k_thread *thread);
+
+void sys_trace_thread_cancel(struct k_thread *thread);
+
+void sys_trace_thread_suspend(struct k_thread *thread);
+
+void sys_trace_thread_resume(struct k_thread *thread);
+
+void sys_trace_thread_ready(struct k_thread *thread);
+
+void sys_trace_thread_pend(struct k_thread *thread);
+
+void sys_trace_thread_info(struct k_thread *thread);
+
+void sys_trace_isr_enter(void);
+
+void sys_trace_isr_exit(void);
+
+void sys_trace_isr_exit_to_scheduler(void);
+
+void sys_trace_idle(void);
+
+void sys_trace_void(unsigned id);
```

Comment:
1. It seems `sys_trace_void` and `sys_trace_end_call` form a sort of start-end tracing pair? That implies the name of `sys_trace_void` should be renamed to `sys_trace_start_void`. And I assume the reasoning behind "void" suffix is that it does not really take any arguments other than ID -- which allows it to be matched with its end.
2. I assume other `sys_trace_start_`{`void`,`int`,`char`,etc.} would eventually appear?
3. Reflecting my point 2 above, this adhoc-ID seems like an anti-pattern, whereas the other tracing functions (like `sys_trace_thread_ready`) above encode back-end agnostically the same meaning. Consider `sys_trace_start_void(SYS_TRACE_ID_MUTEX_INIT);` versus `sys_trace_mutex_init();` 
4. Reflecting on my point 3, this would lead to lots of functions. And they are not inline now. For reduced overhead, they should either be inline or simply macros. Macros would be preferrable IMHO as they can access `__LINE__` and `__FUNC__` at callsite and they can decide if they want to call out to a real function. E.g. on Oticon's platform, traces can be single-cycle so a function call incurs significant overhead.

We would happy to help realizing the changes I just mentioned and send a PR.
(Currently I am in progress of adding a new backend for CommonTraceFormat -- that eventually will display nicely in TraceCompass. I am doing this context of native_posix first).

Reply:
> And I assume the reasoning behind "void" suffix is that it does not really take any arguments other than ID -- which allows it to be matched with its end.

void is for empty payload, this is coming from how it is done in systemview, there your have

- SEGGER_SYSVIEW_RecordVoid: Formats and sends a SystemView packet with an empty payload.
- SEGGER_SYSVIEW_RecordU32: Formats and sends a SystemView packet containing a single U32 parameter payload.
- etc.

> I assume other sys_trace_start_{void,int,char,etc.} would eventually appear?

yes, see above.


> Reflecting my point 2 above, this adhoc-ID seems like an anti-pattern, whereas the other tracing functions (like sys_trace_thread_ready) above encode back-end agnostically the same meaning. Consider sys_trace_start_void(SYS_TRACE_ID_MUTEX_INIT); versus sys_trace_mutex_init();

The idea behind IDs is to define and use less APIs, sys_trace_mutex_init will probably only be used in one single place. But, I have to admit, I made it as close as possible to what systemview was expecting and it is easier to implement this way. Someone might choose to add trace points to their own application and the objects defined in their application using existing APIs rather than implement new ones.